azurerm/internal/services/storage/storage_account_resource.go

```diff
@@ -1563,7 +1564,21 @@ func resourceStorageAccountRead(d *pluginsdk.ResourceData, meta interface{}) err
 		d.Set("enable_https_traffic_only", props.EnableHTTPSTrafficOnly)
 		d.Set("is_hns_enabled", props.IsHnsEnabled)
 		d.Set("nfsv3_enabled", props.EnableNfsV3)
-		d.Set("allow_blob_public_access", props.AllowBlobPublicAccess)
+
+		if features.ThreePointOh() {
+			// There is a certain edge case that could result in the Azure API returning a null value for AllowBLobPublicAccess.
+			// Since the field is a pointer, this gets marshalled to a nil value instead of a boolean. Here we set this
+			// to the default value, but since this is a breaking change we'll enforce this in 3.0
+
+			allowBlobPublicAccess := getDefaultAllowBlobPublicAccess()
+			if props.AllowBlobPublicAccess != nil {
+				allowBlobPublicAccess = *props.AllowBlobPublicAccess
+			}
+			d.Set("allow_blob_public_access", allowBlobPublicAccess)
+		} else {
+			d.Set("allow_blob_public_access", *props.AllowBlobPublicAccess)
```

Comment:
the linter would be complaining here, since this can be nil this also avoids a panic:

```suggestion
			d.Set("allow_blob_public_access", props.AllowBlobPublicAccess)
```